### PR TITLE
Fix Event chaining in the on_load event handler return not working

### LINKED ISSUE #610: Event chaining in the on_load event handler return is not working.
**Describe the bug**

Event chaining in the on_load event handler return is not working.

**To Reproduce**
 - Code/Link to Repo:

I implemented it like this

```python
class BaseState(pc.State):
    ...


class QueryState(BaseState):
    label: str

    def init(self):
        print("init!")
        return [
            self.reset_label,
        ]

    def reset_label(self):
        print("reset label!")
        self.label = ""

app = pc.App(state=state.BaseState)
app.add_page(query.index, route="/", on_load=QueryState.init)
```

When I ran the code, the result was as follows.

```
init!
```

The `reset_label()` method doesn't seem to have actually been executed.

**Expected behavior**

The `reset_label()` method should be executed.
This means that the result of its execution should be

```
init!
reset label!
```

**Screenshots**

** Specifics (please complete the following information):**
 - Python Version: 3.11.0
 - Pynecone Version: 0.1.18
 - OS: Mac OS Monterey 12.6
 - Browser (Optional): .
 
**Additional context**

### PR DESCRIPTION
**Pull request by ElijahAhianyo**
This PR fixes a couple of issues with the way on-load events are handled. In this PR, the following should work now:

- event-chaining( handlers that return other handlers)
- list of on-load events (eg. `app.add_page('index', on_load = [State.first_handler, State.second_handler]`)
- Async handlers

fixes #610 , #603

### CODE REVIEW

File: pynecone/app.py [hunk 1]
```diff
@@ -148,7 +148,7 @@ def add_cors(self):
             allow_origins=["*"],
         )
 
-    def preprocess(self, state: State, event: Event) -> Optional[Delta]:
+    async def preprocess(self, state: State, event: Event) -> Optional[Delta]:
```
Comment (picklelo) [inline pynecone/app.py:151]: Should we make `postprocess` async also for consistency?

File: pynecone/app.py [hunk 2]
```diff
@@ -400,7 +400,7 @@ def compile(self, force_compile: bool = False):
 
 async def process(
     app: App, event: Event, sid: str, headers: Dict, client_ip: str
-) -> StateUpdate:
+) -> Union[StateUpdate, List[StateUpdate]]:
```
Comment (picklelo) [inline pynecone/app.py:403]: We can just always return `List[StateUpdate]` to simplify things

File: pynecone/app.py [hunk 3]
```diff
@@ -430,21 +430,27 @@ async def process(
     state.router_data[constants.RouteVar.CLIENT_IP] = client_ip
 
     # Preprocess the event.
-    pre = app.preprocess(state, event)
-    if pre is not None:
+    pre = await app.preprocess(state, event)
+    if pre is not None and not isinstance(pre, List):
         return StateUpdate(delta=pre)
```
Comment (picklelo) [inline pynecone/app.py:435]: We can listify this here

File: pynecone/app.py [hunk 4]
```diff
@@ -582,7 +588,13 @@ async def on_event(self, sid, data):
         update = await process(self.app, event, sid, headers, client_ip)
 
         # Emit the event.
-        await self.emit(str(constants.SocketEvent.EVENT), update.json(), to=sid)
+        if isinstance(update, List):
```
Comment (picklelo) [inline pynecone/app.py:591]: We can make sure update is always a List so we don't need this check

File: pynecone/middleware/hydrate_middleware.py [hunk 5]
```diff
@@ -37,20 +39,42 @@ def preprocess(self, app: App, state: State, event: Event) -> Optional[Delta]:
...
+                    load_event = [load_event]
+                updates = []
+                for single_event in load_event:
+                    updates.append(
+                        await self.execute_load_event(
+                            state, single_event, event.token, event.payload
+                        )
+                    )
+                return updates
             return format.format_state({state.get_name(): state.dict()})
```
Comment (picklelo) [inline pynecone/middleware/hydrate_middleware.py:52]: We can maybe just make this a `StateUpdate` object to simplify the API

File: pynecone/middleware/hydrate_middleware.py [hunk 6]
```diff
@@ -37,20 +39,42 @@ def preprocess(self, app: App, state: State, event: Event) -> Optional[Delta]:
...
+            ValueError: If the state value is None.
         """
         substate_path = format.format_event_handler(load_event).split(".")
         ex_state = state.get_substate(substate_path[:-1])
-        load_event.fn(ex_state)
+        if ex_state:
+            return await state.process_event(
+                handler=load_event, state=ex_state, payload=payload, token=token
+            )
+        else:
```
Comment (picklelo) [inline pynecone/middleware/hydrate_middleware.py:77]: Don't need the `else` since the first case returns

File: pynecone/state.py [hunk 7]
```diff
@@ -546,37 +546,65 @@ def get_substate(self, path: Sequence[str]) -> Optional[State]:
...
         """
         # Get the event handler.
         path = event.name.split(".")
         path, name = path[:-1], path[-1]
         substate = self.get_substate(path)
         handler = substate.event_handlers[name]  # type: ignore
 
-        # Process the event.
-        fn = functools.partial(handler.fn, substate)
+        if substate:
```
Comment (picklelo) [inline pynecone/state.py:566]: We can change this to

```python
if not substate:
    raise ValueError(...)

return ...

```

File: pynecone/middleware/hydrate_middleware.py [hunk 8]
```diff
@@ -37,20 +39,42 @@ def preprocess(self, app: App, state: State, event: Event) -> Optional[Delta]:
...
-    def execute_load_event(self, state: State, load_event: EventHandler) -> None:
+    async def execute_load_event(
+        self, state: State, load_event: EventHandler, token: str, payload: Dict
+    ) -> StateUpdate:
         """Execute single load event.
 
         Args:
             state: The client state.
             load_event: A single load event to execute.
+            token: client token
```
Comment (picklelo) [inline pynecone/middleware/hydrate_middleware.py:62]: nit: can we format the comments with uppercase and periods like "A client token." and "The event payload."

File: pynecone/state.py [hunk 9]
```diff
@@ -546,37 +546,65 @@ def get_substate(self, path: Sequence[str]) -> Optional[State]:
...
+            )
+
+    async def process_event(
+        self, handler: EventHandler, state: State, payload: Dict, token: str
+    ) -> StateUpdate:
+        """Process event.
+
+        Args:
+            handler: Eventhandler to process.
+            state: state to process the handler.
```
Comment (picklelo) [inline pynecone/state.py:585]: Same nit about formatting the arg comments

File: pynecone/state.py [hunk 10]
```diff
@@ -546,37 +546,65 @@ def get_substate(self, path: Sequence[str]) -> Optional[State]:
...
+                state=substate,
+                payload=event.payload,
+                token=event.token,
+            )
+        else:
+            raise ValueError(
+                "The value of state cannot be None when processing an event."
+            )
+
+    async def process_event(
```
Comment (picklelo) [inline pynecone/state.py:578]: Nice way to abstract this!

**Review by picklelo — COMMENTED** [review]
Nice work! Mostly minor comments

File: pynecone/app.py (hunk 1 above)
Comment (ElijahAhianyo) [reply pynecone/app.py:151]: yep, I agree

**Review by picklelo — APPROVED** [review]
thanks for the updates!

**Pull request merged**